### PR TITLE
Add support for _embedded_ PKCS7 signing to signtool_signer sign()

### PR DESCRIPTION
**Pull request by out0xb2**
To the sign() function:
* added type hints
* In parameter dict "signature_options", the key "sign_alg" is
deprecated and replaced with "format".  Its prior value "pkcs12" was
NOT a signature format, it was the format of the signer's
key file. Moving it to a new signer_option key "key_file_format"
* New "signature_option" key "format" supports 1 value of "pkcs7" which
is the true signature format being produced
* Added "format_options" key to parameter dict "signature_options". Values
"embedded" and "detachedSignatureData" are added to distinguish the
different PKCS7 flavors supported.  Note that these format_options
are mutually exclusive
* To ensure compatibility with existing code that passes
"signature_options" with key "pkcs12", sign() detects this and
fills the new parameters with values the provide backwards compatible
behavior

### CODE REVIEW

**Comment by out0xb2** [comment]
All unit tests in edk2toolext\tests\capsule\* pass (with a couple skipped test cases

File: edk2toolext/capsule/signtool_signer.py [hunk 1]
```diff
@@ -20,6 +20,10 @@
 from edk2toollib.utility_functions import RunCmd
 
 GLOBAL_SIGNTOOL_PATH = None
+SUPPORTED_SIGNTOOL_FORMAT_OPTIONS = {
```
Comment (out0xb2) [inline edk2toolext/capsule/signtool_signer.py:23]: I should probably rename SIGNTOOL to SIGNATURE... as I think this was supposed to be a signature_options['format_options'] ...

File: edk2toolext/capsule/signtool_signer.py [hunk 2]
```diff
@@ -36,18 +40,40 @@ def get_signtool_path():
...
-    # NOTE: Currently, we only support the necessary algorithms for capsules.
-    if signature_options['sign_alg'] != 'pkcs12':
-        raise ValueError(f"Unsupported signature algorithm: {signature_options['sign_alg']}!")
+
+    # NOTE: Currently, we only support the necessary options for capsules & Windows Firmware Policies
+
+    # The following _if_ clause handles the deprecated signature_option 'sign_alg' for backwards compatibility
+    # when the deprecated option is supplied, this code adds the new, required options based on prior code behavior
+    if 'sign_alg' in signature_options:
+        print('!!! WARNING: signature_option "sign_alg" is deprecated, use "format" instead')
```
[the remark this answers is not in the document]
Comment (out0xb2) [reply edk2toolext/capsule/signtool_signer.py:54]: [Python logging HOWTO](https://docs.python.org/3/howto/logging.html#when-to-use-logging) advises:

> Best tool for the task ...
> warnings.warn() in library code if the issue is avoidable and the client application should be modified to eliminate the warning
> logging.warning() if there is nothing the client application can do about the situation, but the event should still be noted

[the remark this answers is not in the document]
Comment (corthon) [reply edk2toolext/capsule/signtool_signer.py:54]: Neat!
I guess I'd go with "warnings.warn".

[the remark this answers is not in the document]
Comment (out0xb2) [reply edk2toolext/capsule/signtool_signer.py:54]: Pending commit manual test output:
```
>python -W default::DeprecationWarning create_policy.py

Calling signtool ...
c:\git\edk2-pytool-extensions\edk2toolext\capsule\signtool_signer.py:54: DeprecationWarning: Signature_option "sign_alg" is deprecated, use "format"
  warnings.warn('Signature_option "sign_alg" is deprecated, use "format"', DeprecationWarning)
signtool complete.
```

File: edk2toolext/capsule/signtool_signer.py (hunk 1 above)
Comment (out0xb2) [reply edk2toolext/capsule/signtool_signer.py]: Addressed in commit 2

File: edk2toolext/capsule/signtool_signer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (out0xb2) [reply edk2toolext/capsule/signtool_signer.py]: Addressed in commit 2

File: edk2toolext/capsule/pyopenssl_signer.py [hunk 3]
```diff
@@ -21,10 +22,30 @@ def sign(data, signature_options, signer_options):
...
+            signer_options['key_file_format'] = 'pkcs12'
+        else:
+            raise ValueError(f"Unsupported signature algorithm: {signature_options['sign_alg']}!")
+
+    ''' signature type 'bare' is just a signed digest, no headers/footers or ASN '''
+    if signature_options['type'] != 'bare':
+        raise ValueError(f"Unsupported signature type: {signature_options['type']}")
+    if 'type_options' in signature_options:
+        raise ValueError(f"Signature type options not supported")
+    if signature_options['encoding'] != 'binary':
```
Comment (corthon) [inline edk2toolext/capsule/pyopenssl_signer.py:43]: Binary in this case is PEM?

File: edk2toolext/capsule/pyopenssl_signer.py [hunk 4]
```diff
@@ -21,10 +22,30 @@ def sign(data, signature_options, signer_options):
     dictionaries that are used by capsule_tool and capsule_helper
```
Comment (corthon) [inline edk2toolext/capsule/pyopenssl_signer.py:22]: Might as well add the type annotations/hints. ^^^

File: edk2toolext/capsule/pyopenssl_signer.py (hunk 3 above)
Comment (out0xb2) [reply edk2toolext/capsule/pyopenssl_signer.py:43]: As I understand it, PEM is base64 with headers and footers.
This is the raw, binary signed digest... followed by 0x0A... no other headers or footers.

Comment (out0xb2) [reply edk2toolext/capsule/pyopenssl_signer.py:43]: Scratch to 0x0A, that was being added by an editor.
See line 38 docstring

Comment (out0xb2) [reply edk2toolext/capsule/pyopenssl_signer.py:43]: Resolved in latest commit

File: edk2toolext/capsule/pyopenssl_signer.py (hunk 4 above)
Comment (out0xb2) [reply edk2toolext/capsule/pyopenssl_signer.py:22]: Added, tested all configurations, no breaks (of my code anyway).

**Comment by out0xb2** [comment]
> Looks good on the whole. Just added a question and a suggestion.

Updated per your feedback, all unit tests and my test scenario pass (I'm now testing all 4 configuration, OpenSSL, signtool, new parameters, and legacy parameters)

**Comment by out0xb2** [comment]
whoops, not the button I was looking for...

**Comment by out0xb2** [comment]
@corthon , awaiting your, or other admin with write access, to approve

File: edk2toolext/capsule/pyopenssl_signer.py (hunk 4 above)
Comment (out0xb2) [reply edk2toolext/capsule/pyopenssl_signer.py:22]: Passed unit tests as well, resolved in commit d16b0339e4caaab3d6b7f0bd18027f8b52b34cf2

**Pull request merged**